File: ferrolearn-python/tests/divergence_knn_multichunk_parallel_on_y.py
"""Divergence pin: `ferrolearn.KNeighborsClassifier` diverges from the DEFAULT
multi-core `sklearn.neighbors.KNeighborsClassifier` in the BRUTE regime for
`n_samples > 256` on exact distance ties.

Audit of the #2143 default-backend fix
(`ferrolearn-neighbors/src/knn.rs` `brute_parallel_on_y` + `find_neighbors`).
The fix ports sklearn's `parallel_on_Y` reduction
(`sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp`) as a SINGLE
per-query max-heap filled over ALL `n` training points in global index order,
then a second pass through that heap's slots
(`_parallel_on_Y_synchronize`-style double heap).

That single-global-heap model is only correct when sklearn uses ONE
chunk-thread, i.e. `n_samples <= pairwise_dist_chunk_size` (256) OR
`OMP_NUM_THREADS=1`. For `n_samples > 256` sklearn splits Y into
`Y_n_chunks = ceil(n / 256)` chunks and, on any machine with >= 2 hardware
threads (the universal default), builds ONE heap PER chunk-thread and merges
them in `_parallel_on_Y_synchronize`
(`_argkmin.pyx.tp:237-261`, thread_num-major then jdx). That per-chunk-then-
merge reduction resolves an exact distance tie at the k-th boundary differently
from a single global-index-order heap.

The builder's `find_neighbors` doc-comment claims the brute path "matches the
live sklearn 1.5.2 oracle ... bit-for-bit" and asserts (for `n < 256`) "there is
exactly one Y-chunk and one thread, so the result is fully deterministic" — true
for `n < 256`, but the implementation extends that single-heap model to
`n > 256`, where sklearn's default is the multi-chunk merge. ferrolearn returns
the `OMP_NUM_THREADS=1` tie resolution that a default multi-core sklearn user
never sees.

Verification model B (goal.md): the expected value is the LIVE sklearn 1.5.2
oracle computed in-test (default thread pool — the multi-core contract); nothing
is literal-copied from the ferrolearn side (R-CHAR-3).

Tracking: #2.
"""

import numpy as np

import ferrolearn
from sklearn.neighbors import KNeighborsClassifier as SkKNN


def _fixture():
    """Explicit (no-RNG) brute-regime fixture, n=400 > 256 -> 2 Y-chunks.

    X is all zero except column 0 = 10.0, with four "sure winner" overrides at
    indices 5/250/260/390 (distances 1/2/3/4 from the origin query). Every other
    training point is at distance exactly 10.0 — a large exact tie, of which
    k - 4 members fill the neighbor list. n_features = 20 > 15 forces `brute`
    (`sklearn/neighbors/_base.py:610-618`).
    """
    n, d = 400, 20
    X = np.zeros((n, d), dtype=np.float64)
    X[:, 0] = 10.0
    for idx, val in [(5, 1.0), (250, 2.0), (260, 3.0), (390, 4.0)]:
        X[idx, 0] = val
    # Labels: distinct classes on the two members the backends disagree on, so
    # the tie-SET difference is observable in predict_proba. sklearn keeps idx 3
    # (class 1), ferrolearn keeps idx 0 (class 0); the six shared neighbors are
    # class 2.
    y = np.full(n, 2, dtype=np.int64)
    y[0] = 0
    y[3] = 1
    Q = np.zeros((1, d), dtype=np.float64)
    return X, y, Q


def test_divergence_multichunk_predict_matches_sklearn():
    """sklearn DEFAULT (multi-core) brute classifier at n=400>256 produces a
    full class-probability vector that the n>256 tie SET determines; this guards
    the observable downstream of the kneighbors tie SET via the binding's
    public surface (`predict` / `predict_proba`; `kneighbors` is not bound).

    `predict` (argmax) happens to agree here (the six shared neighbors dominate),
    so this asserts the full neighbor-derived class counts via `predict_proba` in
    the companion test; here we additionally assert ferrolearn's predicted label
    equals sklearn's so the pin documents the argmax-stable / proba-flipped
    split explicitly."""
    X, y, Q = _fixture()
    k = 8

    sk = SkKNN(n_neighbors=k).fit(X, y)
    assert sk._fit_method == "brute"
    sk_pred = sk.predict(Q)

    fl = ferrolearn.KNeighborsClassifier(n_neighbors=k).fit(X, y)
    fl_pred = np.asarray(fl.predict(Q))

    # predict argmax is class 2 for both (the 6 shared neighbors); the divergence
    # is in the *probabilities* (companion test). This assert documents that the
    # label is the stable part of the split.
    np.testing.assert_array_equal(fl_pred, sk_pred)


def test_divergence_multichunk_predict_proba_flip():
    """The tie-SET difference flips `predict_proba` between ferrolearn and the
    DEFAULT multi-core sklearn classifier."""
    X, y, Q = _fixture()
    k = 8

    sk = SkKNN(n_neighbors=k).fit(X, y)
    sk_proba = sk.predict_proba(Q)

    fl = ferrolearn.KNeighborsClassifier(n_neighbors=k).fit(X, y)
    fl_proba = np.asarray(fl.predict_proba(Q))

    np.testing.assert_allclose(
        fl_proba,
        sk_proba,
        rtol=0,
        atol=0,
        err_msg=(
            "predict_proba diverges: sklearn DEFAULT multi-core keeps tie member "
            f"idx 3 (class 1) -> {sk_proba.tolist()}; ferrolearn keeps idx 0 "
            f"(class 0) -> {fl_proba.tolist()}. The n>256 brute parallel_on_Y "
            "multi-chunk tie resolution differs from ferrolearn's single-heap "
            "(OMP_NUM_THREADS=1) resolution."
        ),
    )

File: ferrolearn-neighbors/tests/divergence_knn_multichunk_parallel_on_y.rs
//! Divergence pin: ferrolearn's brute-regime `kneighbors` reproduces sklearn's
//! `parallel_on_Y` reduction with a SINGLE per-query max-heap built over ALL `n`
//! training points in index order. That is only correct for sklearn's
//! `chunks_n_threads == 1` (single Y-chunk OR `OMP_NUM_THREADS=1`). For
//! `n_samples > pairwise_dist_chunk_size` (256) sklearn splits Y into multiple
//! chunks and, on any machine with >= 2 hardware threads, builds one max-heap
//! PER chunk-thread and merges them in `_parallel_on_Y_synchronize`
//! (`sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp:237-261`).
//! That per-chunk-then-merge reduction resolves an exact distance tie at the
//! k-th boundary DIFFERENTLY from a single global-index-order heap.
//!
//! This is the DEFAULT user-facing output: a default `KNeighborsClassifier`
//! (`metric='minkowski'`, `p=2`, `algorithm='auto'`) with `n_features > 15`
//! routes to `brute` (`sklearn/neighbors/_base.py:610-618`), whose auto
//! strategy is `parallel_on_Y` here, and on a normal multi-core box that is the
//! output every user sees — it is NOT an `OMP_NUM_THREADS=1` corner case.
//!
//! ## Live sklearn 1.5.2 / numpy 2.4.5 oracle (system python3), this fixture
//!
//! ```text
//! n=400, d=20, k=8.  X = zeros((400,20)); X[:,0] = 10.0 ;
//!   X[5,0]=1 ; X[250,0]=2 ; X[260,0]=3 ; X[390,0]=4 .   Q = zeros((1,20)).
//! y = arange(400) % 2 .   KNeighborsClassifier(n_neighbors=8).fit(X,y)
//!   ._fit_method                         -> 'brute'
//!   .kneighbors(Q, return_distance=True)[1][0]:
//!
//!   OMP_NUM_THREADS=1   -> [5, 250, 260, 390, 2, 4, 0, 6]
//!   OMP_NUM_THREADS>=2  -> [5, 250, 260, 390, 6, 2, 4, 3]   (STABLE for 2..28)
//! ```
//!
//! For `n=400`, `Y_n_chunks = ceil(400/256) = 2`, so `chunks_n_threads =
//! min(2, effective_n_threads)`. Hence the `>=2` value is the SAME for every
//! thread count from 2 to 28 (verified) — it is a stable, reproducible,
//! deterministic contract for any default multi-core sklearn install, NOT
//! thread-noise. The four points at index {0,2,3,4,6} are all at distance 10.0
//! (an exact tie); sklearn keeps the multiset {6,2,4,3}, ferrolearn keeps
//! {2,4,0,6}. The first four neighbors (the strict winners 5/250/260/390) and
//! all eight distances are identical; only the tie SET + ORDER at the k-boundary
//! diverge.
//!
//! ferrolearn (`brute_parallel_on_y`, single global-index-order heap,
//! `ferrolearn-neighbors/src/knn.rs`) returns the `OMP_NUM_THREADS=1` value
//! `[5, 250, 260, 390, 2, 4, 0, 6]`, which a default multi-core sklearn user
//! never sees. The builder's doc-comment claims the brute path "matches the
//! live sklearn 1.5.2 oracle ... bit-for-bit" and that for `n < 256` "there is
//! exactly one Y-chunk and one thread, so the result is fully deterministic" —
//! true, but the implementation extends that SINGLE-heap model to `n > 256`,
//! where sklearn's default is the MULTI-chunk merge.
//!
//! Expected values are the LIVE sklearn oracle, regenerated fresh, NOT copied
//! from ferrolearn (R-CHAR-3). Tracking: #2144.

use ferrolearn_core::traits::Fit;
use ferrolearn_neighbors::{Algorithm, KNeighborsClassifier};
use ndarray::{Array1, Array2};

/// Explicit (no-RNG) reconstruction of the oracle fixture.
///
/// `X`: 400x20, all zero except column 0 = 10.0, with four "sure winner"
/// overrides at indices 5/250/260/390 (distances 1/2/3/4 from the origin
/// query). Every other training point is at distance exactly 10.0 — a 396-way
/// exact tie, of which `k - 4 = 4` members fill the neighbor list.
fn fixture() -> (Array2<f64>, Array1<usize>, Array2<f64>) {
    let n = 400usize;
    let d = 20usize;
    let mut x = Array2::<f64>::zeros((n, d));
    for i in 0..n {
        x[[i, 0]] = 10.0;
    }
    for (idx, val) in [(5usize, 1.0), (250, 2.0), (260, 3.0), (390, 4.0)] {
        x[[idx, 0]] = val;
    }
    let y = Array1::from_iter(0..n);
    let q = Array2::<f64>::zeros((1, d));
    (x, y, q)
}

/// Divergence: brute-regime (`n_features > 15`) `kneighbors` at `n=400 > 256`
/// returns the `OMP_NUM_THREADS=1` single-heap tie resolution, but sklearn's
/// DEFAULT (multi-core, 2 Y-chunks) `KNeighborsClassifier.kneighbors()` returns
/// a different, stable, deterministic tie resolution.
///
/// sklearn DEFAULT (multi-core) order: [5, 250, 260, 390, 6, 2, 4, 3]
/// ferrolearn (single-heap)     order: [5, 250, 260, 390, 2, 4, 0, 6]
///
/// Tracking: #2144
#[test]
fn divergence_knn_multichunk_parallel_on_y_tie_set() {
    // This fixture (n=400 → 2 Y-chunks) pins the DEFAULT multi-core tie
    // resolution (effective_n_threads >= 2, each chunk its own thread). Under
    // `OMP_NUM_THREADS=1` sklearn itself collapses to the single-heap order
    // `[5, 250, 260, 390, 2, 4, 0, 6]` (which ferrolearn also returns and is
    // equally correct for that env), so skip rather than assert the multi-core
    // expectation in a single-thread environment.
    if std::env::var("OMP_NUM_THREADS").as_deref() == Ok("1") {
        eprintln!("skipped: OMP_NUM_THREADS=1 -> single-heap order (sklearn agrees)");
        return;
    }

    let (x, y, q) = fixture();
    let k = 8usize;

    let fitted = KNeighborsClassifier::<f64>::new()
        .with_n_neighbors(k)
        .with_algorithm(Algorithm::Auto)
        .fit(&x, &y)
        .unwrap();

    let (_dist, indices) = fitted.kneighbors(&q, Some(k)).unwrap();
    let got: Vec<usize> = indices.row(0).to_vec();

    // Live sklearn 1.5.2 / numpy 2.4.5, OMP_NUM_THREADS>=2 (default multi-core),
    // stable across thread counts 2..28:
    //   KNeighborsClassifier(8).fit(X,y).kneighbors(Q)[1][0]
    let sklearn_default_multicore: Vec<usize> = vec![5, 250, 260, 390, 6, 2, 4, 3];

    assert_eq!(
        got, sklearn_default_multicore,
        "n=400>256 brute: ferrolearn kneighbors {got:?} must equal sklearn's \
         DEFAULT multi-core .kneighbors() {sklearn_default_multicore:?} \
         (parallel_on_Y with 2 Y-chunks, stable for OMP_NUM_THREADS 2..28). \
         ferrolearn returns the single-heap / OMP_NUM_THREADS=1 tie resolution \
         [5, 250, 260, 390, 2, 4, 0, 6] instead, which a default multi-core \
         sklearn user never sees."
    );
}
